Refine error messages for partial overrides from `attr_accessor`

jwinr opened the pull request:
This PR modifies Sorbet's handling of overrides where a parent (interface, abstract class, or overridable method) defines only a reader or only a writer, but the child class uses `attr_accessor` (which creates both). We now produce **two** errors if an `attr_accessor :foo` mismatches the parent's definition. One error is for the reader side (`foo`), the other for the writer side (`foo=`). The messages mention "the parent only defines a writer method" (or "reader method"), so users clearly see **why** the override fails and **which** side is problematic. This change also makes sure that we skip override checks for UI-only synthesized methods that don’t represent actual code-level methods (flagged `isAttrBestEffortUIOnly`). 

The PR also includes a small improvement to cache `methodData` and flags, reducing repeated pointer dereferencing and slightly improving readability in the override validation flow.

### Motivation
Fixes #6312 and is a follow-up to address feedback from the previous PR #8400

### Test plan

See included automated tests. This includes `T::Props` (prop/const) tests to be forward-compatible and act as a reminder that the `attr_accessor` logic should apply to prop methods. Currently, this is blocked due to an existing bug that doesn't allow these methods to participate in override checking.

jez commented on test/testdata/definition_validator/attr_accessor_override_error.rb:
```diff
@@ -0,0 +1,102 @@
+  include WriterFooable
+
+  sig {override.returns(String)}
+  attr_accessor :bar # error: Method `WriterOnlyBad#bar=` is marked `override` but the parent only defines a writer method
+# ^^^^^^^^^^^^^^^^^^ error: Method `WriterOnlyBad#bar` is marked `override` but the parent only defines a reader method
```
This is still wrong. We should report two errors, one saying that `bar` doesn't override anything, another saying `bar=` doesn't override anything, but we should not say that "the parent only defines a reader/writer" in either of those error messages. `bar` is unrelated to `foo`.

The fix for this problem would be to get rid of the `override` keyword in the sig for the `attr_accessor`, because the method is unrelated.

jez commented on test/testdata/definition_validator/attr_accessor_override_error.rb:
```diff
@@ -0,0 +1,102 @@
+  include T::Props
+  include PropWritable
+  
+  prop :prop_foo, String, writer: :private
```
What is `writer:`?

jwinr replied:
Sorry for this, looks like a copilot auto-generation (and lesson learned) that I left in by mistake. I'll be removing this as I rework the tests.

jez commented on test/testdata/definition_validator/attr_accessor_override_error.rb:
```diff
@@ -0,0 +1,102 @@
+  include T::Props
+  include PropWritable
+  
+  prop :prop_foo, String # Should error: Method is marked `override` but the parent only defines a writer method
```
The "should error" claim is right, but the reasoning is wrong.

This should error because `prop_foo=` is marked `abstract` in the parent but there is no mention of `override` in the child. That's tracked elsewhere (#1694, #1841, #8282) and is out of scope for this PR.

jez commented on definition_validator/validator.cc:
```diff
@@ -534,9 +534,16 @@ void validateOverriding(const core::Context ctx, const ast::ExpressionPtr &tree,
+    if (overriddenMethods.empty() && methodFlags.isOverride && !methodFlags.isIncompatibleOverride) {
         if (auto e = ctx.state.beginError(methodLoc, core::errors::Resolver::BadMethodOverride)) {
-            e.setHeader("Method `{}` is marked `{}` but does not override anything", methodName, "override");
+            if (methodFlags.isRewriterSynthesized && !ast::MethodDef::Flags().isAttrBestEffortUIOnly) {
```
This `isAttrBestEffortUIOnly` is dead code. You're constructing a `Flags` object using the default constructor, which sets the `isAttrBestEffortUIOnly` field to `false`, then negating what is essentially a constant `false` (`... && !false` is just `...`)

jez replied:
The more I think about this, the less I think looking at `isAttrBestEffortUIOnly` is needed. I think you just need to check whether the other reader/writer method exists.

jwinr replied:
I want to make sure I understand this correctly: Should it display the "the parent only defines a reader/writer method" error if the parent actually has the same base name and the parent has exactly one half (reader or writer) of that name? But in cases where the parent has no matching name at all, or actually has both sides, should we fall back to the generic "does not override anything" error?

jwinr commented:
Sorry for the delay. One of the bigger changes is that I added a `lookupWithoutEq` helper to complement `lookupwithEq`. The idea here was to determine whether the current method is a setter/getter, and then look for the corresponding setter/getter within the superclass chain or the included mixins. So now it would only display the "parent only defines.." error if there is a partial mismatch where the parent defines only the complementary method (reader or writer). We fall back to the "does not override anything" error if there is a total mismatch where the related method doesn’t exist in any ancestor. I also noticed that the existing test for the `Fooable` module was incorrectly claiming that the parent only defines a writer method, when the parent only defines foo (reader), so I’ve corrected this mistake.
